### PR TITLE
Sprint3 showpage

### PR DESCRIPTION
**Pull request by bfl3tch**
- Changes Implemented:
  - Create show pages for when user is not a member:
    - Add "Add Gym" Button
    - Display basic gym details
    - Do not display current gym members
  - Create show pages for when user is a member:
    - Add "Remove Gym" Button
    - Display basic gym details
    - Display my friends who are members.  Next to each friend member, show:
      - Add "Create Event" button (functioning)
      - Add "View Profile" button (functioning)
    - Display members who are not my friend.  Next to each non-friend member, show:
      - Add "Add Friend" button (not working yet)
      - Add "View Profile" button (functioning)
  - Fixed nav bar with ineffective and unnecessary link
  - Updated dashboard to show created gym

- Quality Control Checklist:

  - [x] Code adheres to Rubocop styling (if unavoidable infractions, please clarify below)
  - [ ] 100% SimpleCov test coverage (if below, please clarify below)
    - Test coverage: 89.26%
      - Need to come back and backfill test, if time permits
  - [x] Last commit has passing Travis build

- Blockers (if applicable):

- Next Steps & Additional Notes:
  - Make functioning "Add Friend" button on gym show pages

### CODE REVIEW

File: app/controllers/search_controller.rb [hunk 1]
```diff
@@ -2,4 +2,19 @@ class SearchController < ApplicationController
   def index
     @gyms = BackEndFacade.get_gyms_near_user(params['zip_code'])
   end
+
+  def show
+    @gym = BackEndFacade.get_selected_gym(params[:id])
```
Comment (tvaroglu) [inline app/controllers/search_controller.rb:7]: I saw your note in the PR - I exposed the following route to return users at the same gym that the user is NOT already friends with:

`/users/{user_id}/gym_members?yelp_gym_id={yelp_gym_id}`

This should feed the bottom section of gym show page wireframe (section to add friends)

Comment (tvaroglu) [reply app/controllers/search_controller.rb:7]: For the top section (current friends at the same gym), the existing `/users/{user_id}/friendships` route was refactored to take query param, so without `yelp_gym_id`, you get all user friends (for dashboard), and with query param, you get friends at same gym:

```ruby
  def index
    user = User.find(params[:user_id])
    friends = if params[:yelp_gym_id]
                user.friends_at_same_gym(params[:yelp_gym_id])
              else
                user.followees
              end
     ...
```

Comment (tvaroglu) [reply app/controllers/search_controller.rb:7]: Please let me know if I misinterpreted your question or if you need additional BE support for the required data

File: app/services/back_end_service.rb [hunk 2]
```diff
@@ -34,6 +34,15 @@ def gyms_near_user(zip_code)
       parse_json(response.body)
     end
 
+    def get_one_gym(yelp_gym_id)
+      response = db_conn.get("/api/v1/gym_search/#{yelp_gym_id}")
+      parse_json(response.body)
+    end
+
+    def post_gym_membership(gym_membership_params)
+      db_conn.post("/api/v1/users/#{gym_membership_params[:user_id]}/gym_memberships", gym_membership_params.to_json,'Content-Type' => 'application/json')
```
Comment (tvaroglu) [inline app/services/back_end_service.rb:43]: I would suggest new lines, similar to the `db_conn.post()` method (for readability)

Not important at this point - but we could also abstract the headers to a separate method (`'Content-Type' => 'application/json'`) so we can just call `BackEndService.headers` (or something) to shorten these methods further

File: app/views/search/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<%= @gym.name %><br/>
+<%= @gym.address %><br/>
+<%= @gym.phone %>
+
+<%= form_with url: gyms_path, method: :post, local: true do |f|  %>
+<%= f.hidden_field :user_id, value: current_user.id %>
```
Comment (tvaroglu) [inline app/views/search/show.html.erb:6]: I think if you are passing these as params in the path helper, you don't need the hidden fields too

File: spec/features/gym_search/show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+require 'rails_helper'
+
+RSpec.describe 'gyms show page' do
+  before(:each) do
+    visit gym_path('BJBXzKYxQAXZKb5W6HrRnA')
+  end
+
+  it 'displays the gym and its information', :vcr do
+    expect(page).to have_content("Rishi's Community Yoga")
+    expect(page).to have_content('2750 S Wadsworth Blvd, Bldg D, Ste 202, Denver, CO 80227')
```
Comment (tvaroglu) [inline spec/features/gym_search/show_spec.rb:10]: I would suggest using `expect(page).to have_css `#name` type syntax, and not hard-coding the strings.  Unlikely given cassette's recorded against real data, but still does make for a more fragile test.  If we are getting back data that is rendered into div id's, I would consider that passing (and more dynamic) tests

File: spec/services/back_end_service_spec.rb [hunk 5]
```diff
@@ -41,4 +41,17 @@
...
+    result = BackEndService.get_one_gym('BJBXzKYxQAXZKb5W6HrRnA')
+    expect(result.class).to eq(Hash)
+  end
+
+  it 'can create a gym membership', :vcr do
+    json_response = File.read('./spec/fixtures/gym_membership.json')
+
+    allow(BackEndService).to receive(:post_gym_membership).and_return(JSON.parse(json_response, symbolize_names: true))
+
+    expect(BackEndService.post_gym_membership(json_response).class).to eq(Hash)
```
Comment (tvaroglu) [inline spec/services/back_end_service_spec.rb:55]: Nice work on these!  I was really struggling to get passing (and tested) methods from some of the mocks and stubs

File: spec/fixtures/gym_membership.json [hunk 6]
```diff
@@ -0,0 +1 @@
+{"data":{"id":12,"type":"gym_membership","attributes":{"user_id":5,"yelp_gym_id":"12ic9bhcnd0tr8umv7e1mw","gym_name":"YMCA"}}}
```
Comment (tvaroglu) [inline spec/fixtures/gym_membership.json:1]: Nice, no double-parsing of JSON here ;)

File: app/services/back_end_service.rb (hunk 2 above)
Comment (carolinectan) [reply app/services/back_end_service.rb:43]: Resolved long line. Later refactor for second comment.

File: app/views/search/show.html.erb (hunk 3 above)
Comment (carolinectan) [reply app/views/search/show.html.erb:6]: Removed commented out button that wasn't functional.

File: spec/features/gym_search/show_spec.rb (hunk 4 above)
Comment (carolinectan) [reply spec/features/gym_search/show_spec.rb:10]: Added have_css to testing

**Comment by Scott-Borecki** [comment]
The gym show page for both member and non-member conditions should be working now.  We still need to test this in production to confirm we are functional.

File: app/views/search/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,34 @@
+<div id="gym_attributes">
```
Comment (tvaroglu) [inline app/views/search/show.html.erb:1]: Duplicate of `gyms/show.html.erb`, this should be removed

File: app/controllers/gym_memberships_controller.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class GymMembershipsController < ApplicationController
```
Comment (tvaroglu) [inline app/controllers/gym_memberships_controller.rb:1]: Placeholder - but we need to move `GymsController.destroy` here, and fix the routes.  I would recommend we clean this up proactively as a part of this PR

File: app/controllers/gyms_controller.rb [hunk 9]
```diff
@@ -3,6 +3,22 @@ def index
...
+    @gym_friends = BackEndFacade.get_friends_at_gym(params[:id], current_user.id)
+    @gym_non_friends = BackEndFacade.get_non_friends_at_gym(yelp_gym_id: params[:id], user_id: current_user.id)
+  end
+
+  def create
+    BackEndFacade.create_gym_membership(gym_membership_params)
+    redirect_to dashboard_path(current_user.id)
+  end
+
   def destroy
```
Comment (tvaroglu) [inline app/controllers/gyms_controller.rb:22]: This should go to `GymMemberships` controller

File: app/controllers/search_controller.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+class SearchController < ApplicationController
```
Comment (tvaroglu) [inline app/controllers/search_controller.rb:1]: Do we have a duplicate file here?  This looks like `GymsController`

File: app/facades/back_end_facade.rb [hunk 11]
```diff
@@ -29,6 +29,20 @@ def delete_event(params)
...
     end
 
+    def get_gym_membership(params)
+      response = BackEndService.get_gym_memberships(params[:user_id])
+      return unless response[:data]
+
+      gym_memberships = response[:data].map { |data| GymMembership.new(data) }
+      gym_memberships.select do |gym_membership|
+        gym_membership.yelp_gym_id == params[:yelp_gym_id]
+      end.first
```
Comment (tvaroglu) [inline app/facades/back_end_facade.rb:39]: Brute force for the win

File: app/facades/back_end_facade.rb [hunk 12]
```diff
@@ -60,5 +74,45 @@ def get_user_events(user_id)
...
+
+      YelpGym.new(gym[:data])
+    end
+
+    def get_gym_users(yelp_api_key)
+      users = BackEndService.get_gym_users(yelp_api_key)
+
+      if users[:data]
+        users[:data].map do |user|
+          GymUser.new(user)
```
Comment (tvaroglu) [inline app/facades/back_end_facade.rb]: I would suggest updating the PORO to just be user - but we can come back to for timesake if needed

File: spec/features/gym_search/index_spec.rb [hunk 13]
```diff
@@ -23,4 +23,13 @@
       end
     end
   end
+
+  it 'has a link to each individual gym', :vcr do
```
Comment (tvaroglu) [inline spec/features/gym_search/index_spec.rb:27]: Hardcode makes me nervous, would like to come back to if time (set as a vars in a before block or something)

**Review by carolinectan — APPROVED** [review]

**Review by tvaroglu — APPROVED** [review]
Thank you for the HUGE lift and cleanup  - there were some spots that needed some love (duplicate POROs and whatnot)

File: spec/features/gym_search/index_spec.rb (hunk 13 above)
Comment (Scott-Borecki) [reply spec/features/gym_search/index_spec.rb:27]: I agree!

**Review by Scott-Borecki — APPROVED** [review]
This looks good to me as long as Travis builds!

**Pull request merged**